snapshot/bundlestore/groupcache.go

```diff
@@ -0,0 +1,93 @@
+package bundlestore
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+
+	"github.com/golang/groupcache"
+	"github.com/scootdev/scoot/cloud/cluster"
+)
+
+type PeerFetcher interface {
+	Fetch() ([]cluster.Node, error)
+}
+
+type GroupcacheConfig struct {
+	Name         string
+	Memory_bytes int64
+	AddrSelf     string
+	Endpoint     string
+	Fetcher      PeerFetcher
+}
+
+// Add in-memory caching to the given store.
+func MakeGroupcacheStore(underlying Store, cfg *GroupcacheConfig) (Store, http.Handler, error) {
+	// Fetch peers and create a list of addresses including self.
+	peerNodes, err := cfg.Fetcher.Fetch()
+	if err != nil {
+		return nil, nil, err
+	}
+	peers := []string{}
+	for _, node := range peerNodes {
+		peers = append(peers, string(node.Id()))
+	}
+	log.Print("Adding groupcacheStore peers: ", peers)
+
+	// Create and initialize peer group.
+	// The HTTPPool constructor will register as a global PeerPicker on our behalf.
+	poolOpts := &groupcache.HTTPPoolOptions{BasePath: cfg.Endpoint}
+	pool := groupcache.NewHTTPPoolOpts(cfg.AddrSelf, poolOpts)
+	pool.Set(peers...)
+
+	// Create the cache which knows how to retreive the underlying bundle data.
+	var cache = groupcache.NewGroup(cfg.Name, cfg.Memory_bytes, groupcache.GetterFunc(
+		func(ctx groupcache.Context, bundleName string, dest groupcache.Sink) error {
+			log.Print("Not cached, fetching bundle: ", bundleName)
+			if reader, err := underlying.OpenForRead(bundleName); err != nil {
```

Comment:
This if/else if/else block reads strangely - is this a Go thing? It's not obvious right away that the successful code path should execute OpenForRead then ReadAll then SetBytes. Unless this is some idiom I would rather have the admittedly verbose if err != nil checks broken out.